core/shared/src/main/scala/zio/ZRef.scala

```diff
@@ -15,182 +15,107 @@
-   * @tparam C type of the value of the `Ref` to be modified
-   * @return `IO[E, C]` modified value of the `Ref`
-   */
-  def modifySome[E, C](default: C)(pf: PartialFunction[B, (C, A)])(implicit ev1: EA <:< E, ev2: EB <:< E): IO[E, C]
-
-  /**
-   * Writes a new value to the `Ref`, with a guarantee of immediate
-   * consistency (at some cost to performance).
-   *
-   * @param a value to be written to the `Ref`
-   * @return `IO[EA, Unit]`
-   */
+  def fold[EC, ED, C, D](
+    ea: EA => EC,
+    eb: EB => ED,
+    ca: C => Either[EC, A],
+    bc: B => Either[ED, D]
+  ): ZRef[EC, ED, C, D]
+
   def set(a: A): IO[EA, Unit]
 
-  /**
-   * Writes a new value to the `Ref` without providing a guarantee of
-   * immediate consistency.
-   *
-   * @param a value to be written to the `Ref`
-   * @return `IO[EA, Unit]`
-   */
   def setAsync(a: A): IO[EA, Unit]
 
-  /**
-   * Atomically modifies the `Ref` with the specified function. This is not
-   * implemented in terms of `modify` purely for performance reasons.
-   *
-   * @param f function to atomically modify the `Ref`
-   * @return `IO[E, Unit]`
-   */
-  def update[E](f: B => A)(implicit ev1: EA <:< E, ev2: EB <:< E): IO[E, Unit] =
-    modify(b => ((), f(b)))
-
-  /**
-   * Atomically modifies the `Ref` with the specified function and returns the
-   * updated value. This is not implemented in terms of `modify` purely for
-   * performance reasons.
-   *
-   * @param f function to atomically modify the `Ref`
-   * @return `IO[E, B]` modified value of the `Ref`
-   */
-  def updateAndGet[E](f: B => A)(implicit ev1: EA <:< E, ev2: EB <:< E): IO[E, B] =
-    modify(b => (b, f(b)))
-
-  /**
-   * Atomically modifies the `Ref` with the specified partial function.
-   * If the function is undefined on the current value it doesn't change it.
-   *
-   * @param pf partial function to atomically modify the `Ref`
-   * @return `IO[E, Unit]`
-   */
-  def updateSome[E](pf: PartialFunction[B, A])(implicit ev1: EA <:< E, ev2: EB <:< E): IO[E, Unit]
-
-  /**
-   * Atomically modifies the `Ref` with the specified partial function.
-   * If the function is undefined on the current value it returns the old value
-   * without changing it.
-   *
-   * @param pf partial function to atomically modify the `Ref`
-   * @return `IO[E, B]` modified value of the `Ref`
-   */
-  def updateSomeAndGet[E](pf: PartialFunction[B, A])(implicit ev1: EA <:< E, ev2: EB <:< E): IO[E, B]
+  final def collect[C](pf: PartialFunction[B, C]): ZRef[EA, Option[EB], A, C] =
+    fold(identity, Some(_), Right(_), pf.lift(_).fold[Either[Option[EB], C]](Left(None))(Right(_)))
+
+  final def contramap[C](f: C => A): ZRef[EA, EB, C, B] =
+    contramapEither(c => Right(f(c)))
+
+  final def contramapEither[EC >: EA, C](f: C => Either[EC, A]): ZRef[EC, EB, C, B] =
+    dimapEither(f, Right(_))
+
+  final def dimap[C, D](f: C => A, g: B => D): ZRef[EA, EB, C, D] =
```

Comment:
`mapBoth` / `mapBothError`? Maybe a little strained because one is contramapping but then again, dimap doesn't draw any such distinction.

Reply:
We can. I thought `dimap` was the name we used for this function, for example on `Schedule` and `Sink`?

Reply:
Ah, ok, in that case, let's stick with `dimap`.